api/v1beta1/plugin_list.go

```diff
@@ -35,100 +40,101 @@ func (p GrafanaPlugin) String() string {
-func (l PluginList) Update(plugin *GrafanaPlugin) {
-	for i, installedPlugin := range l {
-		if installedPlugin.Name == plugin.Name {
-			l[i].Version = plugin.Version
-			break
-		}
+	if listedVersion.Compare(requestedVersion) == -1 {
+		p.Version = version
+
+		return
 	}
 }
 
-// Sanitize remove duplicates and enforce semver
-func (l PluginList) Sanitize() PluginList {
-	var sanitized PluginList
+// Helps to simplify version consolidation
+type PluginMap map[string]GrafanaPlugin
 
-	for _, plugin := range l {
-		if plugin.HasInvalidVersion() {
+func (m PluginMap) Merge(plugins PluginList) {
+	for _, p := range plugins {
+		if p.HasInvalidVersion() {
 			continue
 		}
 
-		if !sanitized.HasSomeVersionOf(&plugin) {
-			sanitized = append(sanitized, plugin)
+		if plugin, ok := m[p.Name]; ok {
+			plugin.Update(p.Version)
+			m[p.Name] = plugin
+		} else {
+			m[p.Name] = p
 		}
 	}
+}
 
-	return sanitized
+func (m PluginMap) GetPluginList() PluginList {
+	return slices.Collect(maps.Values(m))
 }
 
-// HasSomeVersionOf returns true if the list contains the same plugin in the exact or a different version
-func (l PluginList) HasSomeVersionOf(plugin *GrafanaPlugin) bool {
-	for _, listedPlugin := range l {
-		if listedPlugin.Name == plugin.Name {
-			return true
-		}
-	}
+func NewPluginMap() PluginMap {
+	pm := PluginMap{}
 
-	return false
+	return pm
 }
 
-// HasExactVersionOf returns true if the list contains the same plugin in the same version
-func (l PluginList) HasExactVersionOf(plugin *GrafanaPlugin) bool {
-	for _, listedPlugin := range l {
-		if listedPlugin.Name == plugin.Name && listedPlugin.Version == plugin.Version {
-			return true
-		}
-	}
+func NewPluginMapFromList(plugins PluginList) PluginMap {
+	pm := PluginMap{}
+
+	pm.Merge(plugins)
 
-	return false
+	return pm
 }
 
-// HasNewerVersionOf returns true if the list contains the same plugin but in a newer version
-func (l PluginList) HasNewerVersionOf(plugin *GrafanaPlugin) (bool, error) {
-	for _, listedPlugin := range l {
-		if listedPlugin.Name != plugin.Name {
-			continue
-		}
+type PluginList []GrafanaPlugin
 
-		if listedPlugin.Version == PluginVersionLatest && plugin.Version != PluginVersionLatest {
-			return true, nil
-		}
+func (l PluginList) String() string {
+	plugins := make(sort.StringSlice, 0, len(l))
 
-		if plugin.Version == PluginVersionLatest {
-			return false, nil
-		}
+	for _, plugin := range l {
+		plugins = append(plugins, plugin.String())
+	}
 
-		listedVersion, err := semver.Parse(listedPlugin.Version)
-		if err != nil {
-			return false, err
-		}
+	sort.Sort(plugins)
 
-		requestedVersion, err := semver.Parse(plugin.Version)
-		if err != nil {
-			return false, err
-		}
+	return strings.Join(plugins, ",")
+}
 
-		if listedVersion.Compare(requestedVersion) == 1 {
-			return true, nil
-		}
-	}
+// Sanitize remove duplicates and enforce semver
+func (l PluginList) Sanitize() PluginList {
+	plugins := NewPluginMapFromList(l)
 
```

Comment:
I think this needs to return a sorted list, otherwise ConfigMap updates can generate some diffs.